flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/sink/KafkaCommitter.java

```diff
@@ -61,13 +62,25 @@ private void commitTransaction(KafkaCommittable committable) {
                 new FlinkKafkaInternalProducer<>(createKafkaProducerConfig(transactionalId))) {
             producer.resumeTransaction(committable.getProducerId(), committable.getEpoch());
             producer.commitTransaction();
-        } catch (InvalidTxnStateException | ProducerFencedException e) {
-            // That means we have committed this transaction before.
+        } catch (InvalidTxnStateException e) {
             LOG.warn(
-                    "Encountered error {} while recovering transaction {}. "
-                            + "Presumably this transaction has been already committed before",
-                    e,
-                    committable);
+                    "Unable to commit recovered transaction ({}) because it's in invalid state. "
+                            + "Most likely the transaction has been aborted for some reason. Please check Kafka logs for more details.",
+                    committable,
+                    e);
+        } catch (ProducerFencedException e) {
+            LOG.warn(
+                    "Unable to commit recovered transaction ({}) because its producer is already fenced."
+                            + " This means that you either have a different producer with the same '{}' (this is"
+                            + " unlikely with the '{}' as all generated ids are unique and shouldn't be reused)"
+                            + " or recovery took longer than '{}' ({}ms). In both cases this most likely signals data loss,"
+                            + " please consult documentation for more details.",
```

Comment:
I assume the Flink docs? (here, it's a bit ambiguous whether you mean the Kafka or Flink docs)
```suggestion
                            + " please consult the Flink documentation for more details.",
```